proposals/Proposal-1-Time-Partitioner-Refactoring.md

```diff
@@ -0,0 +1,176 @@
+
+### Partitioner
+
+The partitioner is utilized to determine the distribution or organization of messages when flushed to cloud storage.
+The Cloud Storage sink connector offers two partitioners:
+
+- **Simple partitioner**: This is the default partitioning method based on Pulsar partitions. In other words, data is
+  partitioned according to the pre-existing partitions in Pulsar topics. For instance, a message for the
+  topic `public/default/my-topic-partition-0` would be directed to the
+  file `public/default/my-topic-partition-0/xxx.json`, where `xxx` signifies the message offset.
+
+- **Time partitioner**: Data is partitioned according to the time it was flushed. Using the previous message as an
+  example, if it was received on 2023-12-20, it would be directed
+  to `public/default/my-topic-partition-0/2023-12-20/xxx.json`, where `xxx` also denotes the message offset.
+
+## Motivation
+
+- Incorrect implementation of time partitioner: The current implementation only adds time information to the file path
+  , while both the Simple Partitioner and Time Partitioner partition messages based on the topic partition.
+  The Time Partitioner is merely a special case of the Simple Partitioner. The existing time partitioner does not
+  actually partition messages based on time.
+
+- Inflexible current partitioner: It's hard for now to implement a correct Time Partitioner
+  based on the current partitioner interface. The current connector first splits messages based on the topic, then
+  allows the partitioner to
+  generate the file path. As a result, all current partitioner implementations are based on the topic.
+
+- Non-intuitive Partitioner interface:
+  The [existing Partitioner interface](https://github.com/streamnative/pulsar-io-cloud-storage/blob/master/src/main/java/org/apache/pulsar/io/jcloud/partitioner/Partitioner.java)
+  is not user-friendly. It has four methods, but it actually don't need so many methods. For instance, the expected
+  behavior of `encodePartition` and `generatePartitionedPath` is overlapped.
+  We should make its implementation simple and clear enough.
+
+## Goals
+
+### In Scope
+
+- Refactor the existing partitioner implementation to improve its intuitiveness and ease of use.
+- Introduce a new partitioner interface that includes two partitioners: Topic Partitioner and Time Partitioner.
+- Ensure backward compatibility to avoid disrupting current partitioner usage.
+
+### Out of Scope
+
+- None
+
+## High Level Design
+
+A new `Partitioner` interface will be added, with two partitioner implementations: `TopicPartitioner`
+and `TimePartitioner`.
+
+The behavior of these partitioners is as follows:
+
+- **Topic Partitioner**: Messages are partitioned according to the pre-existing partitions in the Pulsar topics. For
+  instance, a message for the topic `public/default/my-topic-partition-0` would be directed to the
+  file `public/default/my-topic-partition-0/xxx.json`, where `xxx` signifies the message offset.
```

Comment:
> where `xxx` signifies the message offset.

Is this the offset that represents the earliest or the last offset?

Reply:
The earliest. I have also updated the proposal to make it clearer.